Adafruit_MQTT.cpp

```diff
@@ -297,13 +297,13 @@ bool Adafruit_MQTT::disconnect() {
 }
 
 
-bool Adafruit_MQTT::publish(const char *topic, const char *data, uint8_t qos) {
-    return publish(topic, (uint8_t*)(data), strlen(data), qos);
+bool Adafruit_MQTT::publish(const char *topic, const char *data, uint8_t qos, bool retain) {
```

Comment:
I'd like the avoid changing function signatures, is there a way to update this method without changing the signature?

Reply:
I set the retain parameter to default to false in the header definition so I believe this should be backwards compatible with existing clients. Please correct me if that assumption is false though as I don't have a ton of experience with C++

Reply:
I've done a bunch of c++, so let me chime in on this one. ;)
I see two options here:
1) make **retain** an optional parameter :  ```retain=false```
2) create a new method, so the existing one remains undisturbed

See [this link](https://stackoverflow.com/questions/3784114/how-to-pass-optional-arguments-to-a-method-in-c) for more info on option 1

Option 2 may be the cleanest, so the caller does not get confused about the implicit.
What that means is that you would add a new public function called something like:
```
bool Adafruit_MQTT::publishWithRetain(const char *topic, const char *data, uint8_t qos, bool retain) {
// original code goes here
}

bool Adafruit_MQTT::publish(const char *topic, const char *data, uint8_t qos) {
  // all this function does now is to call the new one, with the default reatin value
  return publishWithRetain(topic, data, qos, false /*retain*/);
}
```
Hope this makes sense

Reply:
Update: after looking that the existing code, it seems to me that a better approach for this
may be to simply [overload](https://www.geeksforgeeks.org/operator-overloading-c/) publish, as it is currently done for publish that does not use "uint16_t bLen".
Sorry for not seeing that sooner. ;)

In other words:

```c++
  // Publish a message to a topic using the specified QoS level.  Returns true
  // if the message was published, false otherwise.
  bool publish(const char *topic, const char *payload, uint8_t qos = 0);
  bool publish(const char *topic, uint8_t *payload, uint16_t bLen, uint8_t qos = 0);
  bool publish(const char *topic, /*const*/ uint8_t *payload, uint16_t bLen, uint8_t qos, bool retain);   // ADD THIS
```

```c++
bool Adafruit_MQTT::publish(const char *topic, const char *data, uint8_t qos) {
  return publish(topic, (uint8_t*)(data), strlen(data), qos, false);
}

bool Adafruit_MQTT::publish(const char *topic, uint8_t *data, uint16_t bLen, uint8_t qos) {
  return publish(topic, data, bLen, qos, false);
}

bool Adafruit_MQTT::publish(const char *topic, /*const*/ uint8_t *data, uint16_t bLen, uint8_t qos, bool retain) {
  // Construct and send publish packet.
  uint16_t len = publishPacket(buffer, (uint16_t) sizeof(buffer), topic, data, bLen, qos /*, ADD retain HERE...*/);
  if (!sendPacket(buffer, len))
    return false;
...
```